### PR TITLE
Add RhythmFormer

### PR DESCRIPTION
**Pull request by gauthier-th**
This PR adds the RhythmFormer model to rPPG-Toolbox.

From [RhythmFormer: Extracting rPPG Signals Based on Hierarchical Temporal Periodic Transformer](https://doi.org/10.48550/arXiv.2402.12788), by Zou *et al.*, 2024

### CODE REVIEW

File: configs/infer_configs/PURE_MMPD_RHYTHMFORMER_BASIC.yaml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        DO_CROP_FACE: True
+        USE_LARGE_FACE_BOX: True
+        LARGE_BOX_COEF: 1.5
+        DETECTION:
+          DO_DYNAMIC_DETECTION: False
+          DYNAMIC_DETECTION_FREQUENCY : 30
+          USE_MEDIAN_FACE_BOX: False    # This should be used ONLY if dynamic detection is used
+      RESIZE:
+        H: 128
+        W: 128
```
Comment (yahskapar) [inline configs/infer_configs/PURE_MMPD_RHYTHMFORMER_BASIC.yaml:62]: Since these are inference configs and the `TRAIN` and `VALID` configurations won't really be used, can you remove them from this file and in other infer config files?

File: configs/infer_configs/UBFC-rPPG_PURE_RHYTHMFORMER.yaml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  DROP_RATE: 0.2
+  NAME: RhythmFormer
+INFERENCE:
+  BATCH_SIZE: 4
+  EVALUATION_METHOD: "FFT"        # "FFT" or "peak detection"
+  EVALUATION_WINDOW:
+    USE_SMALLER_WINDOW: False        # Change this if you'd like an evaluation window smaller than the test video length
+    WINDOW_SIZE: 10        # In seconds
+  #MODEL_PATH:   "PreTrainedModels/UBFC_cross_RhythmFormer.pth"
+  MODEL_PATH:   "/run/user/1000/gvfs/sftp:host=imvia-6-168.pu6306.rp.u-bourgogne.fr,user=gauthier/home/gauthier/PreTrainedModels/UBFC-rPPG_SizeW128_SizeH128_ClipLength160_DataTypeStandardized_DataAugNone_LabelTypeStandardized_Crop_faceTrue_Large_boxTrue_Large_size1.5_Dyamic_DetFalse_det_len30_Median_face_boxFalse/UBFC_PURE_RhythmFormer_Epoch19.pth"
```
Comment (yahskapar) [inline configs/infer_configs/UBFC-rPPG_PURE_RHYTHMFORMER.yaml:104]: Update this `MODEL_PATH` - seems specific to your machine when it should be similar to the other model paths (i.e., relative).

File: neural_methods/loss/RythmFormerLossCompuer.py
Comment (yahskapar) [inline neural_methods/loss/RythmFormerLossCompuer.py:1]: Correct the naming of this file, looks like it should be "Computer" instead of "Compuer".

Comment (yahskapar) [reply neural_methods/loss/RythmFormerLossCompuer.py:1]: May also want to consolidate some of these loss functions in the future instead of a separate file for each model, but this can be discussed later.

File: neural_methods/model/RhythmFormer.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+        x = self.Fusion_Stem(x)    #[N*D 64 H/4 W/4]
+        x = x.view(N,D,64,H//4,W//4).permute(0,2,1,3,4)
+        x = self.patch_embedding(x)    #[N 64 D 8 8]
+        for i in range(3):
+            x = self.stages[i](x)    #[N 64 D 8 8]
+        features_last = torch.mean(x,3)    #[N, 64, D, 8]  
+        features_last = torch.mean(features_last,3)    #[N, 64, D]  
+        rPPG = self.ConvBlockLast(features_last)    #[N, 1, D]
+        rPPG = rPPG.squeeze(1)
+        return rPPG
```
Comment (yahskapar) [inline neural_methods/model/RhythmFormer.py:418]: Add a newline to the end of this file.

File: neural_methods/trainer/RhythmFormerTrainer.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            calculate_metrics(predictions, labels, self.config)
+
+
+    def save_model(self, index):
+        if not os.path.exists(self.model_dir):
+            os.makedirs(self.model_dir)
+        model_path = os.path.join(
+            self.model_dir, self.model_file_name + '_Epoch' + str(index) + '.pth')
+        torch.save(self.model.state_dict(), model_path)
+        print('Saved Model Path: ', model_path)
```
Comment (yahskapar) [inline neural_methods/trainer/RhythmFormerTrainer.py:176]: Add a newline to the end of this file.

File: neural_methods/loss/RythmFormerLossCompuer.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    @staticmethod
+    def HR_loss(inputs, target,  diff_flag , Fs, std):
+        psd_gt, psd_pred, SNR, macc = calculate_metric_per_video(inputs.detach().cpu(), target.detach().cpu(), diff_flag = diff_flag, fs=Fs, hr_method='Peak')
+        pred_distribution = [normal_sampling(np.argmax(psd_pred), i, std) for i in range(psd_pred.size)]
+        pred_distribution = [i if i > 1e-15 else 1e-15 for i in pred_distribution]
+        pred_distribution = torch.Tensor(pred_distribution).to(torch.device('cuda'))
+        target_distribution = [normal_sampling(np.argmax(psd_gt), i, std) for i in range(psd_gt.size)]
+        target_distribution = [i if i > 1e-15 else 1e-15 for i in target_distribution]
+        target_distribution = torch.Tensor(target_distribution).to(torch.device('cuda'))
+        return kl_loss(pred_distribution , target_distribution)
```
Comment (yahskapar) [inline neural_methods/loss/RythmFormerLossCompuer.py:167]: Add a newline to the end of this file.

**Review by yahskapar — CHANGES_REQUESTED** [review]
Minor changes requested. Otherwise, everything looks good to me. Thank you for this PR!

File: configs/infer_configs/UBFC-rPPG_PURE_RHYTHMFORMER.yaml (hunk 2 above)
Comment (gauthier-th) [reply configs/infer_configs/UBFC-rPPG_PURE_RHYTHMFORMER.yaml:104]: Oops, forgot to change this one

**Review by yahskapar — APPROVED** [review]
LGTM now, thanks again for your PR!
